tls.c

```diff
@@ -0,0 +1,194 @@
+#include "memcached.h"
+
+#ifdef TLS
+
+#include "tls.h"
+#include <string.h>
+#include <sysexits.h>
+
+static pthread_mutex_t ssl_ctx_lock = PTHREAD_MUTEX_INITIALIZER;
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+
+void SSL_LOCK() {
+    pthread_mutex_lock(&(ssl_ctx_lock));
+}
+
+void SSL_UNLOCK(void) {
+    pthread_mutex_unlock(&(ssl_ctx_lock));
+}
+
+/*
+ * Reads decrypted data from the underlying BIO read buffers,
+ * which reads from the socket.
+ */
+ssize_t ssl_read(conn *c, void *buf, size_t count) {
+    assert (c != NULL);
+    /* TODO : document the state machine interactions for SSL_read with
+        non-blocking sockets/ SSL re-negotiations
+    */
+    return SSL_read(c->ssl, buf, count);
+}
+
+/*
+ * SSL sendmsg implementation. Perform a SSL_write.
+ */
+ssize_t ssl_sendmsg(conn *c, struct msghdr *msg, int flags) {
+    assert (c != NULL);
+    size_t bytes, to_copy;
+    int i;
+    bytes = 0;
+    for (i = 0; i < msg->msg_iovlen; ++i)
+        bytes += msg->msg_iov[i].iov_len;
+
+    // ssl_wbuf is pointing to the buffer allocated in the worker thread.
+    assert(c->ssl_wbuf);
+    // TODO: allocate a fix buffer in crawler/logger if they start using
+    // the sendmsg method. Also, set c->ssl_wbuf  when the side thread
+    // start owning the connection and reset the pointer in
+    // conn_worker_readd.
+    // Currntly this connection would not be served by a different thread
+    // than the one it's assigned.
+    assert(c->thread->thread_id == get_thread_id_cb());
+
+    bytes = MIN(bytes, settings.ssl_wbuf_size);
+    to_copy = bytes;
+    char *bp = c->ssl_wbuf;
+    for (i = 0; i < msg->msg_iovlen; ++i) {
+        size_t copy = MIN (to_copy, msg->msg_iov[i].iov_len);
+        memcpy((void*)bp, (void*)msg->msg_iov[i].iov_base, copy);
+        bp +=  copy;
+        to_copy -= copy;
+        if (to_copy == 0)
+            break;
+    }
+    /* TODO : document the state machine interactions for SSL_write with
+        non-blocking sockets/ SSL re-negotiations
+    */
+    return SSL_write(c->ssl, c->ssl_wbuf, bytes);
```

Comment:
Is it really safe to never check SSL for SSL_write()?

Reply:
er, "Is it really safe to never check SSL specific errors for SSL_write()?"

I english good.

Reply:
Not really if we implement the renegotiation. I manually triggered renegotiations at different parts of the state machine and found out the SSL_write is broken as it doesn't return SSL_ERROR_WANT_READ when renegotiation request is in the flight. I suspect this is probably due to the way we set socket flags at the memcached state machine level which interferes with the internal SSL state machine, but still looking through the code to understand how this happens.
This is ok for the moment as I'm disabling the renegotiation, but looks like a bigger change when we want to support it.

Reply:
Yeah lets leave renegotiation disabled. I don't think there's any benefit to it, especially considering these connections should be intra-datacenter and can reconnect quickly. Is that the only time we'd have errors though? I was looking at the code and there're a ton. I can go read more I guess.

Reply:
Not sure about other scenarios except renegotiation, let me check this further.